include/swift/IDE/CompletionInstance.h

```diff
@@ -0,0 +1,97 @@
+//===--- CompletionInstance.h ---------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_IDE_COMPLETIONINSTANCE_H
+#define SWIFT_IDE_COMPLETIONINSTANCE_H
+
+#include "swift/Frontend/Frontend.h"
+#include "llvm/ADT/Hashing.h"
+#include "llvm/ADT/IntrusiveRefCntPtr.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/VirtualFileSystem.h"
+
+namespace swift {
+
+class CompilerInstance;
+class CompilerInvocation;
+class DiagnosticConsumer;
+
+namespace ide {
+
+/// Copy a memory buffer inserting '\0' at the position of \c origBuf.
+std::unique_ptr<llvm::MemoryBuffer>
+makeCodeCompletionMemoryBuffer(const llvm::MemoryBuffer *origBuf,
+                               unsigned &Offset,
+                               llvm::StringRef bufferIdentifier);
+
+/// Manages \c CompilerInstance for completion like operations.
+class CompletionInstance {
+  unsigned MaxASTReuseCount = 100;
+  bool EnableASTCaching = false;
+
+  std::mutex mtx;
+
+  std::unique_ptr<CompilerInstance> CachedCI;
+  llvm::hash_code CachedArgHash;
+  unsigned CachedReuseCound = 0;
+
+  /// Calls \p Callback with cached \c CompilerInstance if it's usable for the
+  /// specified completion request.
+  /// Returns \c if the callback was called. Returns \c false if the
+  /// functionality is disabled, compiler argument has
+  /// changed, primary file is not the same, the \c Offset is not in function
+  /// bodies, or the interface hash of the file has changed.
+  bool performCachedOperaitonIfPossible(
+      const swift::CompilerInvocation &Invocation, llvm::hash_code ArgsHash,
+      llvm::MemoryBuffer *completionBuffer, unsigned int Offset,
+      DiagnosticConsumer *DiagC,
+      llvm::function_ref<void(CompilerInstance &)> Callback);
+
+  /// Calls \p Callback with new \c CompilerInstance for the completion
+  /// request. The \c CompilerInstace passed to the callback already performed
+  /// the first pass.
+  /// Returns \c false if it fails to setup the \c CompilerInstance.
+  bool performNewOperation(
+      swift::CompilerInvocation &Invocation, llvm::hash_code ArgsHash,
+      llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> FileSystem,
+      llvm::MemoryBuffer *completionBuffer, unsigned int Offset,
+      std::string &Error, DiagnosticConsumer *DiagC,
+      llvm::function_ref<void(CompilerInstance &)> Callback);
+
+public:
+  void setEnableASTCaching(bool Flag) { EnableASTCaching = Flag; }
```

Comment:
Even if this were done atomically it would not be correct for racing completions.  Can we pass `EnableASTCaching` in as an argument to `performOperation` so that is is associated with a specific completion?

Reply:
Done! How does it look now?